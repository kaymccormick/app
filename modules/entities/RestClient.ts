import axios from 'axios';
import GetEntitiesResponse from './rest/response/GetEntitiesResponse';
export interface Args {
    baseUri: string;
}

export default class RestClient {
    public baseUri: string;

    public constructor(args: Args) {
        this.baseUri = args.baseUri;
    }
    public getEntities(): Promise<GetEntitiesResponse> {
        return axios.get(`${this.baseUri}/entity`).then(response => {
            if(response.status >= 400) {
                return new Error();
            }
            return response.data;
        }).then(data => {
            const r = new GetEntitiesResponse(data.success, data.result);
            return r;
        });
    }
}
