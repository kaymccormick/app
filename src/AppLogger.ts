export class AppLogger {
debug(message: string, meta?: any) {
let meta2;
if(meta !== undefined) {
  meta2 = Object.assign({}, meta);
delete meta2.msg;
  } else {
  meta2 = {};
  }
let myMsg = message;
if(meta && meta.msg) {
  myMsg += meta.msg;
}
console.log(myMsg);
}
}
