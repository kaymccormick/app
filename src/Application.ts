/** SERVER APPLICATION */
import fs from 'fs';

import path from 'path';
import {Logger} from 'winston';

import { Connection } from 'typeorm';
import {createConnection}from './Factory';
import {ServerConfiguration} from './ServerConfiguration';
import {ServerApplicationModule} from './ServerApplicationModule';

const relModulePath = '../modules';
const moduleRoot = path.join(__dirname, relModulePath);

export class Application {
    /* These are for database entities */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
    public entities: any[] = [];
    public connection?: Connection;
    public get modules(): ServerApplicationModule[] { return this.configuration.modules; }

    public configuration: ServerConfiguration;
    // @ts-ignore
    public logger: Logger;
    public constructor(logger: Logger) {
        this.configuration = new ServerConfiguration();
        this.logger = logger;
        if(logger.debug === undefined) {
            throw new Error('invalid logger');
        }
        this.logger.debug('Application.constructor');
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public init(): Promise<any> {
        fs.readdirSync(moduleRoot).filter((e): boolean => !/^\./.test(e) && fs.statSync(path.join(moduleRoot, e)).isDirectory()).forEach((ent): void => {
            try {
                // eslint-disable-next-line @typescript-eslint/no-var-requires
                const m = require(path.join(relModulePath, ent + '/ServerModule'));
                const serverModule = new m.ServerModule();
                this.configuration.addModule(serverModule);
            } catch(error) {
            }

        });
        this.entities.length = 0;
        this.modules.forEach((module: ServerApplicationModule): void => {
            this.entities.push(...module.getEntities());
        });

        return createConnection(this.entities).then((connection): boolean => {
            this.connection = connection;
            this.modules.forEach((module: ServerApplicationModule): void => {
                module.setup(this, this.configuration);
            });
            return true;
        });
    }

    public getModule(name: string): ServerApplicationModule|undefined {
        return this.modules.find((m): boolean => m.name === name);
    }
}