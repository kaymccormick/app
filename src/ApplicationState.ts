import { CoreModuleKey } from './types';

/**
 * Applicaton state shape. Would like to deprecate this.
 */

export interface ApplicationState {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [moduleKey: string]: any;
}
