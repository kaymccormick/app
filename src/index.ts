export { createConnection } from './Factory';
export { Application } from './Application';
