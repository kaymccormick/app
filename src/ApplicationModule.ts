import { Configuration } from './Configuration';

export interface Actions {
    [actionName: string]: any;
}

export abstract class ApplicationModule {
    public actions: Actions = {};
    public name: string = '';
    public key: string = '';
    public constructor() {
    }
    public abstract setup(config: Configuration): void
    public abstract getReducers(): any;
    public abstract getInitialState(): any;
}

