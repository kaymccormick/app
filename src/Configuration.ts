import { ApplicationModule } from './ApplicationModule';
import { Resources,Reducers }from './types';
export class Configuration {
    protected resources: Resources = {};
    public modules: ApplicationModule<any>[] = [];
    public addModule(module: ApplicationModule<any>) {
        this.modules.push(module);
    }
    public addResource(resourceName: string, resource: any): void {
        this.resources[resourceName] = resource;
    }
    public getInitialState(): any {
        const initialState: any = {};
        this.modules.forEach(module => initialState[module.key] = module.getInitialState());
        return initialState;
    }

    public collectReducers() {
        const reducers: Reducers = {};
        this.modules.forEach(module => {
            const r = module.getReducers();
            reducers[module.key] = r;
        });
        return reducers;
    }
    public getResource(name: string): any {
        return this.resources[name];
    }

}
