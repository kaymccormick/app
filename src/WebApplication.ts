import {Configuration} from './Configuration';
import {ApplicationModule} from './ApplicationModule';
import {Module as ClassModelModule } from '../modules/classModel';
import {Module as EntitiesModule } from '../modules/entities';
import { Store, applyMiddleware, createStore, compose, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { ApplicationState,Action } from '../model/types';

const modules = [];
const relModulePath = '../modules';

export class WebApplication {
    public store?: Store<ApplicationState, Action>;
    public get modules(): ApplicationModule<any>[] { return this.configuration.modules; }

    public configuration: Configuration;
    public constructor() {
        this.configuration = new Configuration();
    }

    public handleChange() {
    }

    public init(): void {
    /* This is totaly gross! */
        const m = new ClassModelModule();
        m.setup(this, this.configuration);
        this.configuration.addModule(m);
        const m2 = new EntitiesModule();
        m2.setup(this, this.configuration);
        this.configuration.addModule(m2);

        const reducers = this.configuration.collectReducers();
        // @ts-ignore
        this.store = createStore(combineReducers(reducers), compose(applyMiddleware(thunk),  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()));
        this.store!.subscribe(this.handleChange);
    }


}
