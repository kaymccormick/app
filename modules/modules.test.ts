import { Provider } from 'react-redux';
import React from 'react';
import { WebApplication } from '../src/WebApplication';
import { ApplicationModule } from '../src/ApplicationModule';
import { mount, shallow } from 'enzyme';

import RestClient from '@heptet/rest-client';

let restClient= new RestClient({baseUri:'', logDebug: (arg) => console.log(arg)});
let configJs = require('../core.conf').default;
let app = new WebApplication({ config: configJs, restClient });
app.init();

beforeEach(() => {
    configJs = require('../core.conf').default;
    app = new WebApplication({ config: configJs, restClient });
    app.init();
});

test.each(app.modules)('%s', (mod) => {
    expect(mod).toBeDefined();
});

test.each(app.modules)('%s.getInitialState', (mod) => {
    const state = mod.getInitialState();
    expect(state).toBeDefined();
});
test.each(app.modules)('%s.getReducers', (mod) => {
    const reducer = mod.getReducers();
    expect(reducer).toBeDefined();
    expect(typeof reducer).toBe('function');
    const state = reducer(undefined, { type: 'notarealaction' });
    expect(state).toBeDefined();
});
test('unique module keys', () => {
    const keys = app.modules.map((mod) => mod.key);
    expect(app.modules.filter((mod: ApplicationModule<any>, index: number) => keys.indexOf(mod.key, index + 1) !== -1)).toHaveLength(0);
});
test.each(app.modules)('%s.actions', (mod) => {
    const actions = mod.actions;
    Object.keys(actions).forEach((k) => {
        const action = actions[k];
        expect(typeof action).toBe('function');
    });
});
test.each(app.modules)('%s.getMainComponent', (mod) => {
    return mod.getMainComponent().then((MainComponent) => {
        const wrapper = shallow(React.createElement(Provider, { store: app.store! }, React.createElement(MainComponent.default, { module: mod })));
        expect(wrapper.html()).toMatchSnapshot();
    });
});
