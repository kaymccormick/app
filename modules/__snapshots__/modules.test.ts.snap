// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classModel.getMainComponent 1`] = `"<div></div>"`;

exports[`docBrowser.getMainComponent 1`] = `"<div></div>"`;

exports[`editor.getMainComponent 1`] = `"<div><div class=\\"RichEditor-root\\"><button>Debug</button><div class=\\"RichEditor-controls\\"><span class=\\"RichEditor-styleButton\\">H1</span><span class=\\"RichEditor-styleButton\\">H2</span><span class=\\"RichEditor-styleButton\\">H3</span><span class=\\"RichEditor-styleButton\\">H4</span><span class=\\"RichEditor-styleButton\\">H5</span><span class=\\"RichEditor-styleButton\\">H6</span><span class=\\"RichEditor-styleButton\\">Blockquote</span><span class=\\"RichEditor-styleButton\\">UL</span><span class=\\"RichEditor-styleButton\\">OL</span><span class=\\"RichEditor-styleButton\\">Code Block</span></div><div class=\\"RichEditor-controls\\"><span class=\\"RichEditor-styleButton\\">Bold</span><span class=\\"RichEditor-styleButton\\">Italic</span><span class=\\"RichEditor-styleButton\\">Underline</span><span class=\\"RichEditor-styleButton\\">Monospace</span></div><div class=\\"RichEditor-editor\\"><div class=\\"DraftEditor-root\\"><div class=\\"public-DraftEditorPlaceholder-root\\"><div class=\\"public-DraftEditorPlaceholder-inner\\" id=\\"placeholder-e09b4\\" style=\\"white-space:pre-wrap\\">Tell a story...</div></div><div class=\\"DraftEditor-editorContainer\\"><div aria-describedby=\\"placeholder-e09b4\\" class=\\"notranslate public-DraftEditor-content\\" contenteditable=\\"true\\" role=\\"textbox\\" spellcheck=\\"true\\" style=\\"outline:none;user-select:text;-webkit-user-select:text;white-space:pre-wrap;word-wrap:break-word\\"><div data-contents=\\"true\\"><div class=\\"\\" data-block=\\"true\\" data-editor=\\"e09b4\\" data-offset-key=\\"4m8lo-0-0\\"><div data-offset-key=\\"4m8lo-0-0\\" class=\\"public-DraftStyleDefault-block public-DraftStyleDefault-ltr\\"><span data-offset-key=\\"4m8lo-0-0\\"><br data-text=\\"true\\"/></span></div></div></div></div></div></div></div></div></div>"`;

exports[`jsontree.getMainComponent 1`] = `"<div></div>"`;

exports[`logging.getMainComponent 1`] = `"<div></div>"`;

exports[`menus.getMainComponent 1`] = `"<div></div>"`;
